### PR TITLE
Ensure deviatoric of rank two tensor is actually deviatoric

### LINKED ISSUE #15666: Ensure the deviatoric of a rank two tensor is actually deviatoric
## Reason
Sometimes the deviatoric of a tensor can still retain a volumetric component due to numerical rounding.

## Design
Subtract any extra volumetric components to the deviatoric tensor
```
  // Sometimes numerical precision results in a non-zero trace of the deviatoric tensor. Ensure that
  // doesn't happen here by subtracting any extra volumetric components
  if (deviatoric.tr() != 0.0)
    deviatoric.addIa(-1.0 / 3.0 * deviatoric.tr());
```

## Impact
Adds an extra evaluation, but ensures the deviatoric tensor is actually deviatoric.

### PR DESCRIPTION
**Pull request by tophmatthews**
Closes #15666

### CODE REVIEW

**Comment by tophmatthews** [comment]
This fails in bison with an Oe-5 in stress:
```
test:tensor_mechanics/mox_creep.fastmox_creep_therm_only: Running csvdiff: /opt/civet/build_0/moose/python/mooseutils/csvdiff.py /opt/civet/build_0/bison/test/tests/tensor_mechanics/mox_creep/gold/fastmox_creep_therm_only_out.csv /opt/civet/build_0/bison/test/tests/tensor_mechanics/mox_creep/fastmox_creep_therm_only_out.csv --relative-tolerance 5e-05 --abs-zero 1e-08
test:tensor_mechanics/mox_creep.fastmox_creep_therm_only: ERROR: In file fastmox_creep_therm_only_out.csv: The values in column "delta_crp_strain_yy" don't match @ t26
test:tensor_mechanics/mox_creep.fastmox_creep_therm_only: 	relative diff:   2.299e-02 ~ 2.299e-02 = 7.352e-05 (7.352e-05)
test:tensor_mechanics/mox_creep.fastmox_creep_therm_only: In file fastmox_creep_therm_only_out.csv: The values in column "el_strain_yy" don't match @ t26
test:tensor_mechanics/mox_creep.fastmox_creep_therm_only: 	relative diff:   1.500e-04 ~ 1.500e-04 = 6.300e-05 (6.300e-05)
test:tensor_mechanics/mox_creep.fastmox_creep_therm_only: In file fastmox_creep_therm_only_out.csv: The values in column "stress_yy" don't match @ t26
test:tensor_mechanics/mox_creep.fastmox_creep_therm_only: 	relative diff:   3.000e+07 ~ 3.000e+07 = 6.297e-05 (6.297e-05)
```
, but a sign flip in MASTODON:
```
test:materials/isoil.stiffness_and_strength_pressure_dependency: Global variables:
test:materials/isoil.stiffness_and_strength_pressure_dependency:    accel_6x      rel diff:  2.0505044e-09 ~ -6.1326411e-06 = 1.00033e+00 (FAILED)
test:materials/isoil.stiffness_and_strength_pressure_dependency:    accel_6z      rel diff:  1.0489142e-04 ~  1.0489047e-04 = 9.06615e-06 (FAILED)
```

**Comment by tophmatthews** [comment]
Let's move over here @bwspenc @dschwen .

Well I feel like you're ready to overrule me on this one @bwspenc since I can't prove stability improvements beyond the negative porosity values we talked about, but it still seems right that the deviatoric should be as close as possible to a true deviatoric. In most cases, I don't think there'll be a correction or it'll be within machine precision. In some cases, it'll be corrected once. It terrible cases, it'll be twice...

**Comment by hugary1995** [comment]
In the appendix of this paper https://www.sciencedirect.com/science/article/pii/S0045782516311069
The authors propose a correction to the original return map algorithm. In essence, you add a volumetric correction term to ensure the determinant of the plastic flow is 1, which I think is a better way of addressing the issue you have rather than modifying the fundamental deviatoric operator.

**Comment by tophmatthews** [comment]
Wow that's extremely difficult to understand if you don't live in that world...Can you dumb it down for me @hugary1995?

File: framework/include/utils/RankTwoTensorImplementation.h [hunk 1]
```diff
@@ -469,6 +469,19 @@ RankTwoTensorTempl<T>::deviatoric() const
...
+
+  // Sometimes numerical precision results in a non-zero trace of the deviatoric tensor. Ensure that
+  // doesn't happen here by subtracting any extra volumetric components
+  T trace = deviatoric.tr();
+  unsigned int it = 0;
+  while (it < 3 && trace != 0.0)
+  {
+    deviatoric.addIa(-1.0 / 3.0 * trace);
+    trace = deviatoric.tr();
+    if (std::abs(trace) < 1.0e-11)
```
Comment (WilkAndy) [inline framework/include/utils/RankTwoTensorImplementation.h:481]: Just thought i'd comment that the 1.0e-11 is a dimensionful quantity here and might not actually be small.

In my opinion, if your return-map algorithm is not robust enough to handle precision loss then it needs to be re-written.

**Comment by hugary1995** [comment]
Sorry I didn't notice your discussion on slack. I agree with Ben that you'd better address the issue when you can detect negative porosity, e.g. guard your porosity against negative values.

I looked at the line where you calculate the porosity: basically whenever you have zero initial porosity there is a chance you get negative porosity out of it -- that is regardless of the return map algorithm nor the deviatoric operator. Even if you use the fix you are proposing here and it passes the tests, you may still get negative porosity on some other machines.

**Comment by tophmatthews** [comment]
Yes I agree, as I agreed there before. I already have a fix for the negative porosity problem (basically set it back to the initial porosity or some minimum value). I won't have zero initial porosity since it is being range checked at the inputparam stage to be greater than zero and less than one.

This is still bigger than just this one problem though. It results in random (machine error based) coupling between volumetric and J2 inelastic strain terms, which may or may not be minor and below tolerances set for mixing (probably below). But damnit, it just feels wrong to have an operator not doing what it's says it does when there is such an easy fix. I'll take a look more at the paper you suggested, but I'll probably end up closing this as I don't have the time to fight for this.
